### PR TITLE
OptimizeInstructions: Optimize >, >= with an added offset

### PR DESCRIPTION
**Pull request by kripken**
```
x + C1 > C2  ==>  x > C2-C1

e.g.

x + 5 > 100  ==>  x > 95
```

Pretty simple, but we do need to be careful of overflows.

Helps #5008 - this is just `>` and `>=`, and just `+`, so far. See that issue for more
patterns.

Found by the superoptimizer #4994. This was the top suggestion for Java and Dart.

### CODE REVIEW

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -51,6 +51,20 @@
...
+  }
+  switch (type.getBasic()) {
+    case Type::i32:
+      return 32;
+    case Type::i64:
+      return 64;
+    default:
+      return -1;
+  }
+}
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:66]: This is just moved around so it is usable in more places.

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -3567,9 +3574,70 @@ struct OptimizeInstructions
...
+    // give up.
+    auto typeMaxBits = getBitsForType(binary->type);
+    auto leftMaxBits = Bits::getMaxBits(binary->left, this);
+    if (leftMaxBits == typeMaxBits) {
+      return true;
+    }
+    auto rightMaxBits = Bits::getMaxBits(binary->right, this);
+    if (rightMaxBits == typeMaxBits) {
+      return true;
+    }
```
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:3621]: What about use more generic approach?
```ts
if (max(getMaxBits(left), getMaxBits(right)) == typeMaxBits) {
  return true;
}
```

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3621]: I guess we don't use leftMaxBits/rightMaxBits in the code right now, but we do use them in a comment lower down, so I think it's clearer to save them to locals explicitly. But maybe using max would still be slightly shorter, good idea.

**Comment by kripken** [comment]
Actually there may be a better approach here, let me turn this into a draft PR for now.

**Review by tlively — APPROVED** [review]

File: test/lit/passes/optimize-instructions.wast [hunk 3]
```diff
@@ -15111,4 +15111,275 @@
...
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const 0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $gt_u-added-constant (param $x i32)
+    ;; A + 5 > 10  =>  A > 5, since this cannot overflow (A and the constants
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply test/lit/passes/optimize-instructions.wast:15145]: I usually use something like this:
```wat
;; x' + C1 > C2    ->   x' > C2 - C1,  iff x' + C1 doesn't overflow
```

File: src/passes/OptimizeInstructions.cpp [hunk 4]
```diff
@@ -51,6 +51,20 @@
 
 namespace wasm {
 
+static Index getBitsForType(Type type) {
+  if (!type.isBasic()) {
+    return -1;
+  }
+  switch (type.getBasic()) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/OptimizeInstructions.cpp]: Thanks. I changed it to `isNumber()` which seems even clearer here.

**Comment by kripken** [comment]
Feedback addressed, and finished the rewrite I mentioned. This now handles all relational operations on integers at once.

File: test/lit/passes/optimize-instructions.wast [hunk 5]
```diff
@@ -15111,4 +15111,361 @@
...
+  ;; CHECK-NEXT:    (i32.const 0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.ne
+  ;; CHECK-NEXT:    (i32.shr_u
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const -1)
```
Comment (tlively) [inline test/lit/passes/optimize-instructions.wast:15140]: We could further optimize here by seeing that constant -1 has bits set above the maxBits position of the other operand, so they can't possibly be equal.

Comment (kripken) [reply test/lit/passes/optimize-instructions.wast:15140]: Yes, good catch. That is noted here and I'll get to it soon:

https://github.com/WebAssembly/binaryen/issues/5010#issuecomment-1241296045

**Pull request merged**